source/opt/instrument_pass.cpp

```diff
@@ -1008,78 +1008,122 @@ bool InstrumentPass::InstrumentFunction(Function* func, uint32_t stage_idx,
-    if (ecnt == 0)
-      stage = e.GetSingleWordInOperand(kEntryPointExecutionModelInIdx);
-    else if (e.GetSingleWordInOperand(kEntryPointExecutionModelInIdx) !=
-             stage) {
+    stage = e.GetSingleWordInOperand(kEntryPointExecutionModelInIdx);
+    if (stage != SpvExecutionModelVertex &&
+        stage != SpvExecutionModelFragment &&
+        stage != SpvExecutionModelGeometry &&
+        stage != SpvExecutionModelGLCompute &&
+        stage != SpvExecutionModelTessellationControl &&
+        stage != SpvExecutionModelTessellationEvaluation &&
+        stage != SpvExecutionModelTaskNV && stage != SpvExecutionModelMeshNV &&
+        stage != SpvExecutionModelRayGenerationNV &&
+        stage != SpvExecutionModelIntersectionNV &&
+        stage != SpvExecutionModelAnyHitNV &&
+        stage != SpvExecutionModelClosestHitNV &&
+        stage != SpvExecutionModelMissNV &&
+        stage != SpvExecutionModelCallableNV) {
       if (consumer()) {
-        std::string message = "Mixed stage shader module not supported";
+        std::string message = "Stage not supported by instrumentation";
         consumer()(SPV_MSG_ERROR, 0, {0, 0, 0}, message.c_str());
       }
       return false;
     }
-    ++ecnt;
-  }
-  // Check for supported stages
-  if (stage != SpvExecutionModelVertex && stage != SpvExecutionModelFragment &&
-      stage != SpvExecutionModelGeometry &&
-      stage != SpvExecutionModelGLCompute &&
-      stage != SpvExecutionModelTessellationControl &&
-      stage != SpvExecutionModelTessellationEvaluation &&
-      stage != SpvExecutionModelTaskNV && stage != SpvExecutionModelMeshNV &&
-      stage != SpvExecutionModelRayGenerationNV &&
-      stage != SpvExecutionModelIntersectionNV &&
-      stage != SpvExecutionModelAnyHitNV &&
-      stage != SpvExecutionModelClosestHitNV &&
-      stage != SpvExecutionModelMissNV &&
-      stage != SpvExecutionModelCallableNV &&
-      stage != SpvExecutionModelTaskEXT && stage != SpvExecutionModelMeshEXT) {
-    if (consumer()) {
-      std::string message = "Stage not supported by instrumentation";
-      consumer()(SPV_MSG_ERROR, 0, {0, 0, 0}, message.c_str());
-    }
-    return false;
   }
-  // Add together the roots of all entry points
-  std::queue<uint32_t> roots;
+
+  // Next, collect together the roots and stage indices of all entry points
+  bool mixed_stages = false;
+  std::map<uint32_t, std::queue<uint32_t>> stage_to_roots;
   for (auto& e : get_module()->entry_points()) {
-    roots.push(e.GetSingleWordInOperand(kEntryPointFunctionIdInIdx));
+    uint32_t current_stage =
+        e.GetSingleWordInOperand(kEntryPointExecutionModelInIdx);
+    uint32_t current_root =
+        e.GetSingleWordInOperand(kEntryPointFunctionIdInIdx);
+    stage_to_roots[current_stage].push(current_root);
+    if (current_stage != stage) mixed_stages = true;
   }
-  bool modified = InstProcessCallTreeFromRoots(pfn, &roots, stage);
+
+  if (mixed_stages) {
```

Comment:
This function is getting pretty long.  Could the body of this if statement be moved to a function:

```
if (mixed_stages) {
  if (ExistsFunctionCallableWithMultipleShaderStages()) {
     // Emit error
  }
```